src/hotspot/share/runtime/arguments.cpp

```diff
@@ -1029,98 +1023,83 @@ const char* Arguments::handle_aliases_and_deprecation(const char* arg, bool warn
   return NULL;
 }
 
-bool Arguments::parse_argument(const char* arg, JVMFlagOrigin origin) {
-
-  // range of acceptable characters spelled out for portability reasons
-#define NAME_RANGE  "[abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_]"
 #define BUFLEN 255
-  char name[BUFLEN+1];
-  char dummy;
-  const char* real_name;
-  bool warn_if_deprecated = true;
 
-  if (sscanf(arg, "-%" XSTR(BUFLEN) NAME_RANGE "%c", name, &dummy) == 1) {
-    real_name = handle_aliases_and_deprecation(name, warn_if_deprecated);
-    if (real_name == NULL) {
-      return false;
+JVMFlag* Arguments::find_jvm_flag(const char* name, size_t name_length) {
+  char name_copied[BUFLEN+1];
+  if (name[name_length] != 0) {
+    if (name_length > BUFLEN) {
+      return NULL;
+    } else {
+      strncpy(name_copied, name, name_length);
+      name_copied[name_length] = '\0';
+      name = name_copied;
     }
-    JVMFlag* flag = JVMFlag::find_flag(real_name);
-    return set_bool_flag(flag, false, origin);
   }
-  if (sscanf(arg, "+%" XSTR(BUFLEN) NAME_RANGE "%c", name, &dummy) == 1) {
-    real_name = handle_aliases_and_deprecation(name, warn_if_deprecated);
-    if (real_name == NULL) {
-      return false;
+
+  const char* real_name = Arguments::handle_aliases_and_deprecation(name);
+  if (real_name == NULL) {
+    return NULL;
+  }
+  JVMFlag* flag = JVMFlag::find_flag(real_name);
+  return flag;
+}
+
+bool Arguments::parse_argument(const char* arg, JVMFlagOrigin origin) {
+  bool is_bool = false;
+  bool bool_val = false;
+  char c = *arg;
+  if (c == '+' || c == '-') {
+    is_bool = true;
+    bool_val = (c == '+');
+    arg++;
+  }
+
+  const char* name = arg;
+  while (true) {
+    c = *arg;
+    if (isalnum(c) || (c == '_')) {
+      ++arg;
+    } else {
+      break;
     }
-    JVMFlag* flag = JVMFlag::find_flag(real_name);
-    return set_bool_flag(flag, true, origin);
   }
 
-  char punct;
-  if (sscanf(arg, "%" XSTR(BUFLEN) NAME_RANGE "%c", name, &punct) == 2 && punct == '=') {
-    const char* value = strchr(arg, '=') + 1;
+  size_t name_len = size_t(arg - name);
+  JVMFlag* flag = find_jvm_flag(name, name_len);
```

Comment:
Consider -XX:@blah
I think the name_len could be 0 if the preceding while loop did not increment the arg pointer.
But I think it is ok because find_jvm_flag would return NULL on a zero name_len.
Is it correct?

Reply:
The code does handle the case where `name_len==0`, but I added a new check anyway to make the code easier to understand.

Reply:
I tested a few cases and the results are the same before/after adding the `name_len` check.

```
$ java -XX:SharedBaseAddress=0x10000000 --version
java 19-internal 2022-09-20
Java(TM) SE Runtime Environment (slowdebug build 19-internal-adhoc.iklam.ken)
Java HotSpot(TM) 64-Bit Server VM (slowdebug build 19-internal-adhoc.iklam.ken, mixed mode)

$ java -XX:XSharedBaseAddress=0x10000000 --version
Unrecognized VM option 'XSharedBaseAddress=0x10000000'
Did you mean 'SharedBaseAddress=<value>'? Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.

$ java -XX:@SharedBaseAddress=0x10000000 --version
Unrecognized VM option '@SharedBaseAddress=0x10000000'
Did you mean 'SharedBaseAddress=<value>'? Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.

$ java -XX:=0x10000000 --version
Unrecognized VM option '=0x10000000'
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.

$ java -XX:@=0x10000000 --version
Unrecognized VM option '@=0x10000000'
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
```